src/deluge/storage/DX7Cartridge.cpp

```diff
@@ -99,74 +108,115 @@ void DX7Cartridge::packProgram(uint8_t* src, int idx, char* name, char* opSwitch
+ * @return an array of bitfield components as separate bytes
+ */
+template <size_t... args, typename T>
+constexpr std::array<T, sizeof...(args)> unpackBits(T b) {
+	static_assert((args + ...) <= CHAR_BIT);
+	std::array<T, sizeof...(args)> output;
+
+	size_t i = 0;
+	for (size_t num_bits : {args...}) {
+		auto bitmask = static_cast<T>((1 << num_bits) - 1);
+		output[i++] = b & bitmask;
+		b >>= num_bits;
+	}
+	return output;
+}
 
-	return v;
+template <typename T>
+constexpr T clearTopNBits(T b, size_t n) {
+	return b & ~util::top_n_bits<T>(n);
 }
 
-void DX7Cartridge::unpackProgram(uint8_t* unpackPgm, int idx) {
+template <typename T>
+constexpr T clearTopBit(T b) {
+	return clearTopNBits(b, 1);
+}
+
+// For details on the DX7 SYSEX specification, please see contrib/sysex-format.txt
+
+void DX7Cartridge::unpackProgram(std::span<std::uint8_t> unpackPgm, size_t idx) {
 	if (!isCartridge()) {
-		memcpy(unpackPgm, voiceData + 6, 155);
+		std::memcpy(unpackPgm.data(), &voiceData[6], 155);
 		return;
 	}
 
-	// TODO put this in uint8_t :D
-	char* bulk = (char*)voiceData + 6 + (idx * 128);
+	auto* bulk = reinterpret_cast<uint8_t*>(&voiceData[6 + (idx * 128)]);
 
-	for (int op = 0; op < 6; op++) {
-		// eg rate and level, brk pt, depth, scaling
+	for (size_t op = 0; op < 6; ++op) {
+		std::uint8_t* bulk_op = &bulk[op * 17];
+		std::uint8_t* unpack_op = &unpackPgm[op * 21];
 
-		for (int i = 0; i < 11; i++) {
-			uint8_t currparm = bulk[op * 17 + i] & 0x7F; // mask BIT7 (don't care per sysex spec)
-			unpackPgm[op * 21 + i] = normparm(currparm, 99, i);
+		// eg rate and level, brk pt, depth, scaling
+		for (size_t i = 0; i < 11; ++i) {
+			auto currparm = clearTopBit(bulk_op[i]); // mask BIT7 (don't care per sysex spec)
+			unpack_op[i] = normparm(currparm, 99, i);
 		}
 
-		memcpy(unpackPgm + op * 21, bulk + op * 17, 11);
-		char leftrightcurves = bulk[op * 17 + 11] & 0xF; // bits 4-7 don't care per sysex spec
-		unpackPgm[op * 21 + 11] = leftrightcurves & 3;
-		unpackPgm[op * 21 + 12] = (leftrightcurves >> 2) & 3;
-		char detune_rs = bulk[op * 17 + 12] & 0x7F;
-		unpackPgm[op * 21 + 13] = detune_rs & 7;
-		char kvs_ams = bulk[op * 17 + 13] & 0x1F; // bits 5-7 don't care per sysex spec
-		unpackPgm[op * 21 + 14] = kvs_ams & 3;
-		unpackPgm[op * 21 + 15] = (kvs_ams >> 2) & 7;
-		unpackPgm[op * 21 + 16] = bulk[op * 17 + 14] & 0x7F; // output level
-		char fcoarse_mode = bulk[op * 17 + 15] & 0x3F;       // bits 6-7 don't care per sysex spec
-		unpackPgm[op * 21 + 17] = fcoarse_mode & 1;
-		unpackPgm[op * 21 + 18] = (fcoarse_mode >> 1) & 0x1F;
-		unpackPgm[op * 21 + 19] = bulk[op * 17 + 16] & 0x7F; // fine freq
-		unpackPgm[op * 21 + 20] = (detune_rs >> 3) & 0x7F;
+		// FIXME(@stellar-aria): This looks like it's overwriting the previous for loop??? Is this correct?
+		// See original source here:
+		// https://github.com/asb2m10/dexed/blob/a95b61d5196ac018a57ede34bfb037c886551d63/Source/PluginData.cpp#L126
+		std::copy_n(bulk_op, 11, unpack_op);
```

Comment:
yeah it is weird. The DX7 manual indeed does say that 99 is the maximum possible value, so I think the `normparm()` variant is the correct one and the `memcpy()` should be removed.